### PR TITLE
Refactor FXIOS-8401 [v124] - Update message handler to support showing the accessory view login fields are found

### PR DESCRIPTION
**Pull request by nbhasin2**
## :scroll: Tickets
[Jira ticket - 8401](https://mozilla-hub.atlassian.net/browse/FXIOS-8401)
[Github issue - 18628](https://github.com/mozilla-mobile/firefox-ios/issues/18628)

## :bulb: Description

Note: This does not stop the auto password filling but this change will allow us to know when user taps on the username / password fields so that we can show the user accessory view

https://github.com/mozilla-mobile/firefox-ios/assets/8919439/2990adaa-51ee-4a48-a336-41db1c3ee871

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by nbhasin2** [comment]
@issammani Tagging you for the JS portion 🙏🏼

File: firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift:157]: @razvanlitianu

PART1: This is where the magic will happen, once JS side is approved we will use this to know when to show the accessory view. Which when tapped will show the bottom sheet

We need to make small mechanism for this but I can show you or you can look into how other areas of autofill like Addresses / Credit card works. 

See PART2 for how to inject passwords into the page

File: firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift
Comment (nbhasin2) [inline firefox-ios/Client/Frontend/TabContentsScripts/LoginsHelper.swift:361]: @razvanlitianu 
PART2: This will be used to inject password into the tab or webpage you are on

Basically in your bottom sheet as long as you have all the data related to passwords you should be able to use this method to inject it by creating the required object. This also means we will do some cleanup as mentioned in the requestLogins method above and remove that piece. 

Note: There is a chance we might need `currentRequestId:` but will take care of that when needed

**Comment by issammani** [comment]
> @issammani Tagging you for the JS portion 🙏🏼

Nice. Will take a look today 😄

**Review by razvanlitianu — APPROVED** [review]

File: firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js
Comment (issammani) [inline firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js:625]: This has some edge cases that we solve already in desktop that will be tricky if we want to solve them here. Mentioning a few  just to illustrate:
- `input[type="text"], input[type="email"], input:not([type])` selector is not exhaustive. For instance a username can be a `tel` number. 
- Github for instance uses this placeholder for an input to invite collaborators `placeholder="Search by team name, username, full name, or email"` this will be identified wrongly as a username. 
- We also allow here username only forms which is a tricky case to handle and we might have a lot of edge cases with this implementation. We do [quite a few things](https://searchfox.org/mozilla-central/rev/a8cc31504a2379bcf8ba395d2da7bb632b5521d6/toolkit/components/passwordmgr/LoginManagerChild.sys.mjs#868) in desktop to make sure we don't wrongly identify a username only form.

I suggest for now, we handle only forms that have both username and passwords. Then most of the edge cases should be covered by [`inputTypeIsCompatibleWithUsername`](https://github.com/mozilla-mobile/firefox-ios/blob/b5155bed5580c781bb6005d201e3be0553bbc4b6/firefox-ios/Client/Assets/CC_Script/LoginManager.shared.mjs#L10-L16) shared from desktop. The other ones will need better heuristics that we can share later. 

Here is what I am thinking; instead of querying all forms and filtering out the inputs, we should be listening on `focusin` event and then figuring out whether we are in a form or not. Something like this should work and behave the same for the most part:

```js
  function onFocusIn(event) {
    const form = event.target?.form;
    if (!form) {
      return;
    }

    const [username, password] = LoginManagerContent._getFormFields(form, false);
    if (username && password) {
      webkit.messageHandlers.loginsManagerMessageHandler.postMessage({
        type: "fieldType",
        fieldType: event.target === username ? "username" : "password",
      });
    }
  }

  document.addEventListener("focusin", (ev) => onFocusIn(ev));
```
  
  This has a couple of advantages:
  - We don't need to select any forms on load or listen to DOM changes ( also we can now remove that `MutationObserver`). When something is focused we are notified.
  - We don't have to write custom filtering logic. `_getFormFields` takes care of that ( once we share a better heuristic from desktop it would be a drop-in replacement for this)

File: firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js
Comment (issammani) [inline firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js:644]: See my comment above. Attaching focus event listeners on each field might be a bit overkill. Let's flip the logic here. Instead of trying to find the inputs and attaching a `focus` event handler, let's have a `focusin` event on the document and figure out later what was focused ( using `focusin` instead of `focus` because `focus` doesn't bubble )

**Review by issammani — COMMENTED** [review]
I left a couple of comments and suggestions. Let me know what you think 😄

File: firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js
Comment (nbhasin2) [reply firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js]: Oh my, good call
When in future we are doing the work on the platform side then this does seem overkill 😅 
I tested your change locally and it works! Pushing now

File: firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js
Comment (issammani) [inline firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js:594]: Let's allow `password` only forms as well since `_getFormFields` supports it. 
```suggestion
    if (password) {
```

**Review by issammani — APPROVED** [review]
LGTM !

**Comment by nbhasin2** [comment]
image

**Comment by nbhasin2** [comment]
@mergify backport release/v124

**Pull request merged**
